### ISSUE TITLE
De-duping

### ISSUE THREAD

**Issue by bmcfee** [cosmir/openmic-2018#24]
Inspired by a recent tweet on deduping in cifar, I went and performed a similar analysis on our proposed train-test split.  We indeed have some dupes in our dataset, though it seems like they're relatively few.  This does already occur in our (current) train-test split though, so we'll need to do some cleanup to prevent that from happening.

### Procedure

1. Map all tracks to vggish quantized codewords, and flatten the time dimension.  This gives for train a 15000x1280 matrix; for test, 5000x1280.
2. Perform a nearest neighbor search from test to train using L1 distance (though any distance would work).
3. Look for any suspiciously small distances to nearest neighbors.

### Results

![image](https://user-images.githubusercontent.com/1190540/44823187-f2346c00-abcc-11e8-80b4-571877fbf227.png)

This plot shows the distribution of distances from test to 1st, 2nd, ..., 5th nearest neighbor from the training set.  As you can see from the fliers on column 1, there are a few suspiciously low values.  In fact, two of them are identically 0.  The offending culprits are

```
007954_184320 -> 120195_184320
074658_372480 -> 074328_372480
```
Checking the metadata fields, these appear to be exact duplicates with dirty metadata.

There's one more soft near-match, which has distance 9057 (median is about 54000):
```
021334_11520 -> 020898_11520
```
which again, by metadata, seems like a like match.  (The artist ids are identical, but the other metadata is garbage.)

Since in all of these cases, an artist-id filter using metadata would have caught this, I'm wondering if that's sufficient to fix it?  I didn't include this originally because I was under the impression that no artist had more than one track in the collection.  Apparently that assumption was incorrect.

I haven't done a similar analysis on the train-train data, but I suspect there are many more dupes than are caught here.

What do you recommend @ejhumphrey @simondurand ?  I could try to resolve this in the splitter, but I'm not sure how artist conditioning will play with stratified sampling -- in sklearn splitters, we can do either stratification or grouping, but not both simultaneously.  I'll take a look and see if we just got unlucky with this random split, but it'd be nice to have a proper solution in place.

**Comment by ejhumphrey** [comment]
all grumbling aside, yea, I thought we did an artist_id split as well ... guess that's on me.

first, scope of the problem: why not do a full 20k x 20k distance matrix, all splits aside? I'm curious to know what the full duplicate count is. same goes for artist IDs...

second, fixing it: what's the right corrective action here? I see three possibilities, in increasing correctness:
1. leave it, flag it as a known issue, and split the data to avoid train/test contamination
2. drop the dupes and let it be openmic-2018, containing 19980+ examples
3. figure out which positive labels we'd lose for each example, and try to manually backfill from the disjoint set of artists not already covered in the collection.

I think you're suggesting (1) @bmcfee? I don't know which I prefer. (3) sounds like a lot of work, and I think making a point of this at ISMIR is a fine lesson to share.

**Comment by bmcfee** [comment]
> first, scope of the problem: why not do a full 20k x 20k distance matrix, all splits aside? I'm curious to know what the full duplicate count is. same goes for artist IDs...

My laptop only has 8GB ram :cry: 

The number of unique artist ids in the metadata csv is around 7K though, so interpret that however you like.  Since the number of train/test dupes occurs at a rate of around 3 in 5000 (0.06%), I'm not too concerned about it as a general problem, but it'd be nice to have it clean from the get go.

> I think you're suggesting (1) @bmcfee? I don't know which I prefer. (3) sounds like a lot of work, and I think making a point of this at ISMIR is a fine lesson to share.

Indeed, I'd opt for 1.  It'll take a little finessing in the splitter script, but I can try to hack that out tomorrow afternoon/evening.

Opt 2 would also be okay, but aesthetically unpleasing.

As for opt 3, I'm not sure.  The dupes I'm seeing tend to have the same label patterns (though the confidences differ slightly).

**Comment by ejhumphrey** [comment]
okay, well, turns out #6 / #22 can be helpful for this, which is as good as it is bad.

That branch introduces two csvfiles of `sample_key: md5 hash` maps.
* if you look at unique md5 hashes for the audio files, you end up with 20k values.
* if you look at unique md5 hashes for the vggish features, you end up with **19994** values.

Which means, at the VGGish level, you have 6 pairs / 12 samples that are in some sense affected.

This, of course, obviously only catches exact matches ... so the story could be worse for near-misses. I might be able to dig into this a little bit more later, but it's back to real work now with hackdays ending.

**Comment by simondurand** [comment]
I'd vote for 1. And if that is not too difficult to do, also adding an artist conditioning in the stratified sampling seems quite important to me. Even at the expense of a bit less even annotation distribution.

**Comment by bmcfee** [comment]
I agree, artist conditional splits with label stratification would be ideal.  (Duplicate tracks within artist are a nuisance, but they're extremely rare so I can live with it.)  The problem is that the current tools don't support both simultaneously, and it's not (to me) obvious how to achieve that effect.  I'll have some time to think about it this evening.

Alternately, we could brute-force it by rejection-sampling stratified splits until we get one with no cross-contamination.  This would be slow and painful, but in principle should work as long as there's some feasible split out there.

**Comment by simondurand** [comment]
When I was faced with the same problem I just considered groups of tracks from the same artist as one entity but with the length of the number of tracks and used the same method. But it may not work in your label stratification framework. Also, the 15000/5000 split might not work exactly with this method, more something like ~15000/~5000.

If brute force work, go for it. If only a bunch of artists are still in the train and test sets, we can swap them in post processing as it won't significantly change the overall distribution.

**Comment by simondurand** [comment]
I indeed also had a similar problem with this guy. I think he is linked to more than 30k songs on Spotify. Far above anyone else.

I am not entirely sure I understand the difference between 2. and 3.
What I had in mind is:
Let's say you want to assign train/test with this input: 
```
(track_id:'001_001234', piano_yes=1, piano_no=0, violin_yes=0, ..., weight=1)
```
To obtain 
```
(track_id:'001_001234', split:'train', ...) such that sum_{weight \in train} = 15000 and mean_{piano_yes \in train} ~= mean_{piano_yes \in test}, ...
```
Then you do the same with this input: 
```
(artist_id:'12750', piano_yes=A, ..., weight=B)
with A = mean_{track_id \in artist_id:'12750'}(piano_yes) and B = sum_{track_id \in artist_id:'12750'}(weight)
```
To obtain 
```
(artist_id:'12750', split:'train', ...) such that sum_{weight \in train} ~= 15000 and mean_{piano_yes \in train} ~= mean_{piano_yes \in test}, ...
```

I feel like this is not very clear, so do as you think is best ;)

**Comment by bmcfee** [comment]
> I am not entirely sure I understand the difference between 2. and 3.

Option 2 as I understood it would ignore the labels.

Option 3 would apply the same multi-label->majority class reduction logic that we currently use at the track level, but at the artist level.  The idea is as follows:

1. Aggregate the tracks (and annotations) within each artist.  Say the confidence vectors are sum-aggregated.
2. Map each artist id to its maximally confident label.  (Or `_negative` if all associations are negative)
3. Do a stratified split of artists
4. Expand artists back into tracks

There are two main problems that I see here, but I think it's probably our best option:

1. Label diversity within artist could hurt a lot at step 2, especially on rare instruments.
2. Stratification at the level of artists, within high skew of artist->track associations, will imbalance the results at step 3.  We may not be better off than doing a blind split.

**Comment by bmcfee** [comment]
Updates:

I implemented the method described above.  I tried a few random splits, and compared the resulting annotation distributions to the overall population.  Here's an example of resulting probability ratios:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>test_neg</th>
      <th>test_pos</th>
      <th>train_neg</th>
      <th>train_pos</th>
    </tr>
    <tr>
      <th>instrument</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>accordion</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>banjo</th>
      <td>1.0</td>
      <td>0.9</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>bass</th>
      <td>1.0</td>
      <td>1.1</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>cello</th>
      <td>0.9</td>
      <td>1.1</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>clarinet</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>cymbals</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>drums</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>flute</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>guitar</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>mallet_percussion</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>mandolin</th>
      <td>1.0</td>
      <td>1.1</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>organ</th>
      <td>1.1</td>
      <td>0.8</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>piano</th>
      <td>0.9</td>
      <td>1.1</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>saxophone</th>
      <td>0.9</td>
      <td>1.1</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>synthesizer</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>trombone</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>trumpet</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>ukulele</th>
      <td>1.0</td>
      <td>1.1</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>violin</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>voice</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
    <tr>
      <th>_negative</th>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
      <td>1.0</td>
    </tr>
  </tbody>
</table>

In this case, the biggest deviation was `+organ (test)` having ~0.8 probability ratio compared to the full population.  I've seen numbers as low as 0.7, but this gives me an idea.  We can over-sample splits, and specify a maximum allowed deviation for accepted splits.  If we set that number to 0.95, I'm pretty confident that we can get at least one split to come out.  Of course, the sample counts don't pop out to exactly the target 15000/5000, but they're generally pretty close.

I'll do a bit more testing on this, and then push up a new script today.

**Comment by bmcfee** [comment]
After a bit of feature comparison testing, I'm now running into some genuine metadata errors that break artist-conditional filtering.

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>track_id</th>
      <th>album_id</th>
      <th>album_title</th>
      <th>album_url</th>
      <th>artist_id</th>
      <th>artist_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>9352</th>
      <td>69465</td>
      <td>12352</td>
      <td>Classwar Karaoke - 0019 Survey</td>
      <td>http://freemusicarchive.org/music/07_Elements/...</td>
      <td>14230</td>
      <td>07_Elements</td>
    </tr>
    <tr>
      <th>10771</th>
      <td>82089</td>
      <td>13967</td>
      <td>Elements 001-012</td>
      <td>http://freemusicarchive.org/music/Anthony_Dono...</td>
      <td>15896</td>
      <td>Anthony Donovan</td>
    </tr>
  </tbody>
</table>


<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>track_id</th>
      <th>album_id</th>
      <th>album_title</th>
      <th>album_url</th>
      <th>artist_id</th>
      <th>artist_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>10030</th>
      <td>74328</td>
      <td>13126</td>
      <td>slavebation</td>
      <td>http://freemusicarchive.org/music/christian_cu...</td>
      <td>13491</td>
      <td>Artist Name</td>
    </tr>
    <tr>
      <th>10063</th>
      <td>74658</td>
      <td>13178</td>
      <td>Slavebation</td>
      <td>http://freemusicarchive.org/music/Christian_Cu...</td>
      <td>15155</td>
      <td>Christian Cummings</td>
    </tr>
  </tbody>
</table>


<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>track_id</th>
      <th>album_id</th>
      <th>album_title</th>
      <th>album_url</th>
      <th>artist_id</th>
      <th>artist_name</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>13072</th>
      <td>104838</td>
      <td>16424</td>
      <td>Sectioned v4.0</td>
      <td>http://freemusicarchive.org/music/AL355I0/Sect...</td>
      <td>18306</td>
      <td>AL355I0</td>
    </tr>
    <tr>
      <th>12972</th>
      <td>103892</td>
      <td>16322</td>
      <td>Sectioned v4.0</td>
      <td>http://freemusicarchive.org/music/Section_27_N...</td>
      <td>8563</td>
      <td>Section 27 Netlabel</td>
    </tr>
  </tbody>
</table>


It never seems to be more than 3 collisions (out of ~5000 test points), so it's not the end of the world.  Still, we could avoid this by building a full feature hash, and applying some light manual correction to the artist ids prior to the artist-conditional splits.  It still breaks independence assumptions for training/evaluation (which were never really legit to begin with, but c'est la vie), but at least it would rule out obvious dupes.

**Comment by bmcfee** [comment]
Ok, here's a complete list of sample-key feature collisions in the database:
```
{'025401_69120': ['052656_11520'],
 '043546_80640': ['150737_295680'],
 '048491_192000': ['127880_57600'],
 '052656_11520': ['025401_69120'],
 '064241_0': ['073175_119040'],
 '073175_119040': ['064241_0'],
 '075414_76800': ['099644_15360'],
 '099644_15360': ['075414_76800'],
 '127880_57600': ['048491_192000'],
 '146313_15360': ['153411_145920'],
 '150737_295680': ['043546_80640'],
 '153411_145920': ['146313_15360']}
```
The good news is that A) the list is small, and B) the collisions happen only in pairs, so there are no complicated higher-order cliques of colliding tracks to deal with.

At this point, what I'll do is map each colliding sample to the artist id of the lexicographically lower `sample_key`, and use that proxy artist id when generating the train-test split.  This should fix us up going forward.

**Comment by bmcfee** [comment]
Aaaaand the metadata-corrected, pseudo-artist-conditional split seems to do the trick!

![image](https://user-images.githubusercontent.com/1190540/45000403-8ff3b680-af92-11e8-916a-d3f7ada27d91.png)
(Note: no fliers down near the origin on the 1-nn distance distribution.)

Now to script this up, throw in the rejection threshold parameter, and we should be good to go.

**Comment by bmcfee** [comment]
And here's a [link](https://gist.github.com/248a74c7f8ccaad611525ff27677eb16) to the notebook used to generate the sample deduplication index.  I don't think this one needs to be converted into a script (use-once process), but it should be archived along with other development notebooks.

@ejhumphrey @simondurand where would be a good place to store the derived dedupe index?  Alongside metadata / sparse labels?

**Comment by bmcfee** [comment]
New script is implemented and humming along.  I can reliably get a pseudo-artist-conditional split with tolerance of 0.85 (meaning all subsample instrument distributions have 0.85 * p(Y) <= p(Y | train/test) <= p(y) / 0.85), but getting above 0.90 seems to be challenging.  Probably due to JSB.
